### PR TITLE
[RFC] whitespace handling

### PR DESCRIPTION
**Pull request by nvdk**
whitespace handling in the editor is a nuissance and we've had more than our share of bug reports on it. Based on earlier input from @abeforgit I wrote out a possible solution. This needs further refinement, thus the RFC

[rendered](https://github.com/lblod/ember-rdfa-editor/blob/rfc/whitespace-handling/rfcs/whitespace-handling.md)

NOTE: this explicitely doens't cover cursor positioning in empty nodes, which is considered a separate issue. (for more on that see https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_view_filler.html#function-MARKED_NBSP_FILLER

### CODE REVIEW

File: rfcs/whitespace-handling.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+### loading content
+Load content through the insert-html command (and others) becomes more complex because we will want to filter out any unwanted whitespace and linebreaks.
+
+### exporting content
+Exporting content becomes more complex because we will want to translate whitespace that would normally fold (consecutive spaces, linebreaks, spaces at the end of a node, etc.) to something visible.
+
+## Caveats
+### are we the only ones?
+It's unclear whether other editors use this css property and that's something we should investigate. If they don't, perhaps there's a good reason not to do this.
```
Comment (nvdk) [inline rfcs/whitespace-handling.md:60]: https://github.com/ProseMirror/prosemirror/issues/651 seems to confirm prosemirror is also using `white-space: pre-wrap`

**Comment by nvdk** [comment]
one thing to take into account is that this will mean we can't implement a "text-align" feature later on as text-align and whitespace don't work together nicely.

**Comment by abeforgit** [comment]
Nobody wants this more than I do 
I'm excited to investigate conversion strategies, though I'm unsure where to place this in the list of fundamental changes we still need to tackle

**Comment by abeforgit** [comment]
also consider the alignment issue mentioned in your link is from 2018, with any luck this has improved by now

**Comment by nvdk** [comment]
> Nobody wants this more than I do I'm excited to investigate conversion strategies, though I'm unsure where to place this in the list of fundamental changes we still need to tackle

Indeed, I think we will have to improve/bugfix the nbsp handling short term, this is more a long term solution.

File: rfcs/whitespace-handling.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+| prosemirror | use pre-wrap by default, plugins exist to use nbsp behaviour |
+| ckeditor 4 & 5 | use nbsp, seems this is mostly handled in the view |
+| tinymce | use pre-wrap |
+
+
+### old content
+Older documents, pasted content and our templates contain a lot of spurious whitespace which will likely be shown when we use this css property. We will need some kind of strategy to deal with that.
+
+### inconsistencies between layout in the editor and when rendered outside
+Both `pre-wrap` and `break-spaces` also break lines on newline characters, which is likely unwanted behaviour as these will not show up in exported documents unless they apply the same whitespace styling. So extra care needs to be taken to replace these newlines with proper breaks on export.
```
Comment (abeforgit) [inline rfcs/whitespace-handling.md:75]: Likely we will go for inserting <br> elements as a linebreak strategy, so I think this is a non-issue?

**Review by abeforgit — COMMENTED** [review]
from my limited experimentation: 
converting seems doable, but as you discovered we seem to lose some browser-default behavior
which hints at a better handling of dom-updates as a prerequisite for this

**Pull request closed without merging**